lib/rack/lint.rb

```diff
@@ -714,6 +720,28 @@ def verify_content_length(size)
         end
       end
 
+      ## 
+      ## ==== The +rack.protocol+ Header
+      ##
+      def check_rack_protocol_header(status, headers)
+        ## If the +rack.protocol+ header is present, it must be a String, and
+        ## must be one of the values from the +rack.protocols+ array from the
+        ## environment.
+        protocol = headers['rack.protocol']
+        if protocol
+          request_protocols = Array(@env['rack.protocol'])
```

Comment:
If the env entry is required to be an array if present, we shouldn't need to use `Array` to turn it into an array.

Reply:
I'm okay with that, however in HTTP/2, the semantics are that it should only contain a single value, i.e. a string, while in HTTP/1, it can have multiple values. I'm not sure if we want to just specify that it can be a string (single value HTTP/2 semantics) or Array (multiple values HTTP/1 semantics) or an Array with a single value (HTTP/2 wrapped in HTTP/1 semantics). It's probably easier to use if we always assume it's an Array, but 99.9999999999% of the time, this will always be a single value, `websocket`. I've never seen anyone with an upgrade header like `upgrade: websocket, smtp, redis`. I think that's why HTTP/2 standardised it as a single-value pseudo header - realistically no one is asking to upgrade to several distinct protocols, and/or protocol versions. I don't have a strong opinion, but I consider the HTTP/1 semantic to be largely "deprecated" based on the directions the standards appear to be going in. I don't have a strong opinion here either way.

Reply:
If it is going to 99.9999999999% be a string in HTTP/1, and 100% in HTTP/2, we probably should just specify it must be a string.  If a client submits `upgrade: websocket, smtp`, and the app uses the response header `upgrade: websocket`, it wouldn't pass Lint since it wouldn't match, and I think that's OK.  We shouldn't design for the 00.0000000001% use case.

Reply:
I'd be okay with this, but `async-http` will internally follow the HTTP semantics and has it as an Array, which is how servers should be checking it. Therefore, I think we should encourage servers to implement the behaviour correctly, essentially:

```
if env['rack.protocol'].find?{|protocol| protocol.casecmp?('websocket')}
```